code/modules/power/supermatter/supermatter.dm

```diff
@@ -173,130 +233,169 @@
+
 	return 1
 
-/obj/machinery/power/supermatter_shard
 
-/obj/machinery/power/supermatter_shard/bullet_act(obj/item/projectile/Proj)
+/obj/machinery/power/supermatter/bullet_act(var/obj/item/projectile/Proj)
 	var/turf/L = loc
 	if(!istype(L))		// We don't run process() when we are in space
 		return 0	// This stops people from being able to really power up the supermatter
 				// Then bring it inside to explode instantly upon landing on a valid turf.
 
+	if(!has_been_powered)
+		var A = Proj.firer
+		if(ismob(A))
+			var/mob/M = A
+			investigate_log("Projectile was fired by [M.ckey]", "supermatter")
 
-	if(Proj.flag != "bullet")
-		power += Proj.damage * config_bullet_energy
-		if(!has_been_powered)
-			investigate_log("has been powered for the first time.", "supermatter")
-			message_admins("[src] has been powered for the first time <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[x];Y=[y];Z=[z]'>(JMP)</a>.")
-			has_been_powered = 1
+		investigate_log("has been powered for the first time.", "supermatter")
+		message_admins("[src] has been powered for the first time <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[x];Y=[y];Z=[z]'>(JMP)</a>.")
+		has_been_powered = 1
+
+	if(istype(Proj, /obj/item/projectile/beam))
+		power += Proj.damage * config_bullet_energy	* CHARGING_FACTOR / POWER_FACTOR
 	else
 		damage += Proj.damage * config_bullet_energy
 	return 0
 
-/obj/machinery/power/supermatter_shard/singularity_act()
-	var/gain = 100
-	investigate_log("Supermatter shard consumed by singularity.","singulo")
-	message_admins("Singularity has consumed a supermatter shard and can now become stage six.")
-	visible_message("<span class='userdanger'>[src] is consumed by the singularity!</span>")
-	for(var/mob/M in mob_list)
-		M << 'sound/effects/supermatter.ogg' //everyone goan know bout this
-		M << "<span class='boldannounce'>A horrible screeching fills your ears, and a wave of dread washes over you...</span>"
-	qdel(src)
-	return(gain)
-
-/obj/machinery/power/supermatter_shard/attack_paw(mob/user)
-	return attack_hand(user)
-
-
-/obj/machinery/power/supermatter_shard/attack_robot(mob/user)
+/obj/machinery/power/supermatter/attack_robot(mob/user as mob)
 	if(Adjacent(user))
 		return attack_hand(user)
 	else
-		user << "<span class='warning'>You attempt to interface with the control circuits but find they are not connected to your network. Maybe in a future firmware update.</span>"
+		user << "<span class = \"warning\">You attempt to interface with the control circuits but find they are not connected to your network.  Maybe in a future firmware update.</span>"
 	return
 
-/obj/machinery/power/supermatter_shard/attack_ai(mob/user)
-	user << "<span class='warning'>You attempt to interface with the control circuits but find they are not connected to your network. Maybe in a future firmware update.</span>"
+/obj/machinery/power/supermatter/attack_ai(mob/user as mob)
```

Comment:
mob user as mob, the as mob part is useless since you're declaring user as mob already by mob/user